openmdao/drivers/analysis_driver.py

```diff
@@ -0,0 +1,451 @@
+                                   "or specify a number of processors per model that divides "
+                                   f"into {full_size}.")
+
+            color = self._color = comm.rank % size
+            new_comm = comm.Split(color)
+            return new_comm
+
+    def _get_name(self):
+        """
+        Get the name of this DOE driver and its case generator.
+
+        Returns
+        -------
+        str
+            The name of this DOE driver and its case generator.
+        """
+        return self._name
+
+    def run(self):
+        """
+        Generate samples and run the model for each set of generated input values.
+
+        Rank 0 will both manage the distribution of samples to the other procs and
+        serve as a worker running the samples.
+
+        All other procs just run samples.
+
+        Returns
+        -------
+        bool
+            Failure flag; True if failed to converge, False is successful.
+        """
+        comm = self._problem_comm
+        model = self._problem().model
+        self.result.reset()
+        self.iter_count = 0
+
+        # Variables allowed samples are the inputs or implicit outputs in the model.
+        # Non-model-inputs would just have their value overridden when evaluating the model.
+        # Implicit outputs can override the value given in the case, but it might be a
+        # useful mechanism for providing an initial guess.
+        model_inputs = {meta['prom_name'] for _, meta in
+                        model.list_inputs(is_indep_var=True, out_stream=None)}
+        model_implicit_outputs = {meta['prom_name'] for _, meta
+                                  in model.list_outputs(explicit=False, out_stream=None)}
+        self._allowable_vars = model_inputs | model_implicit_outputs
+        n_procs = 1 if comm is None else comm.size
+
+        if self.options['run_parallel'] and MPI and n_procs > 1:
+            batch_size = self.options['batch_size']
+            color_cycler = itertools.cycle(range(self._num_colors))
+            samples_complete = False
+            sample_num = 0
+            job_queues = None
+            colors = comm.gather(self._color, root=0)
+
+            if comm.rank == 0:
+                color_to_rank_map = {num: [i for i, x in enumerate(colors)
+                                     if x == num] for num in set(colors)}
+
+            while not samples_complete:
+                if comm.rank == 0:
+                    job_queues = [deque() for _ in range(n_procs)]
+                    # Rank 0 pushes batch_size jobs to the ranks in job_queues
+                    for i, sample in enumerate(self._samples):
+                        # Ranks of the same color get the same samples
+                        color_idx = next(color_cycler)
+                        for rank_idx in color_to_rank_map[color_idx]:
+                            job_queues[rank_idx].appendleft((sample_num, sample))
```

Comment:
In a later version we might want to revisit how we're distributing the jobs to avoid losing the 'laziness' of some of the generators.